### DISCUSSION TITLE
Custom login page keep adding callbackUrl

### DISCUSSION THREAD

**Discussion by kjugi** [nextauthjs/next-auth#3748] [Help]
# Question

Next-auth ver: 4.1.2

I have configured the custom login route with the credentials provider. It's provided in the `pages` configuration in the `[...nextauth].ts` file. After filling the email and password, I'm using the `signIn` method: https://next-auth.js.org/getting-started/client#signin

```js
signIn('credentials', { ...formData }); // without redirect: false
```

If credentials are good, it's working fine. When I provide wrong creds, the page is reloading and I have got something like this in the url:

```
/login?callbackUrl=http://localhost:3000/login?callbackUrl=http://localhost:3000/&error=CredentialsSignin
```

The first `callbackUrl` looks weird b/c it seem like redirect to the same page. After fail next-auth is reloading the page and it adds another `callbackUrl` param with `error`. I believe that we have some problems here (or I've got the problem).

Problem 1. 
The first `callbackUrl` looks like redirecting to the same page.

Problem 2.
The second `callbackUrl` is wrong b/c it has a callback to the homepage and I thought that I should stay at the login.

Problem 3.
I should have only `error=...` in query params. ~~I was trying to check the error in the `context.req.error` on the server-side but don't have any as in default singIn page.~~ 
> I've checked that and it's taking this prop from `context.req.query` [here](https://github.com/nextauthjs/next-auth/blob/c844296982fec29ff9e9c8e4621cb1d943eda6a6/src/core/index.ts#L77). That part is clear

**What is strange about that behaviour?**

When I'm using the default login page I don't have additional `callbackUrl` query param with the same url and error code. I've got simply redirected to the same site and URL is like this:

`api/auth/signin?error=CredentialsSignin`

I really don't know what I'm missing here in comparison to the default case. I want to get just `error` in query params same as on the default signIn page. Look like the `callbackUrl` is not executed after authorisation fails and it's just reloading the page with adding some other unnecessary `callbackUrl` params.

This part of code looks like responsible for this thing: https://github.com/nextauthjs/next-auth/blob/c844296982fec29ff9e9c8e4621cb1d943eda6a6/src/core/index.ts#L108-L110 

Am I right?

## Summary of the cases
**The cases for the default signIn page:**
- when error, refresh the page and add URL param with the code
- when pass, refresh the page and redirect to the callbackUrl properly

**The cases for custom signIn page (redirect enabled):**
- when error, refreshes the page and add to URL `callbackUrl=...?error=...`
- when pass, refresh the page and stay on login page (don't use callbackUrl)

The cases for the custom signIn page (redirect disabled):
- when error, I need to add my custom validation (that's fine)
- when passing, I can do whatever I want (that's fine too)

# Reproduce

I was trying something on code sandbox but google login doesn't like to work with this site. When provided `NEXTAUTH_URL` env it has some strange problems after login redirect. Anyway here is a demo link:
https://codesandbox.io/s/beautiful-snow-vmwtt?file=/pages/login.js

**Comment by bcomnes** [comment]
I think I'm running into this too. Redirects stopped working when we added a custom login page, apparently because the redirect page gets set to the custom login page, which redirects back to the home page when you are logged in.

**Comment by zanzlender** [comment]
Did you rewrite URL's somewhere in your code? I remember i had a similar issue but with Next middleware, where it would infinitely loop the url because I rewrote the URL based on some statement. It was something with redirect vs rewrite, forgot what exactly... 

A quick alternative for this would be something like this. You stop the redirect and save the response from singIn into a variable. Then check it for error params.

```
  const handleLogin = async (e) => {
    let res = await signIn("credentials", {
      redirect: false,
      password: "12345678",
      email: "email@gmail.com",
    });

    console.log(res);

    --> check for errors and w/e
  };
```

**Comment by DomicaCode** [comment]
Does this have a solution? I seem to keep running into the same issue.

**Comment by antoniojps** [comment]
Having the same issue here.

Want to have a custom login page where I can fetch the error from the query param `error`.

But I noticed that before the `error` parameter I'm always getting the following `callbackUrl` param.
`?callbackUrl=https%3A%2F%2Flocalhost%2Flogin&error=Callback`

I was expecting only this: `?error=Callback`.

So looking into the code, as @kjugi mentioned.
Here we can see the culprit, whenever the user has custom pages it always adds a callbackUrl param regardless.

https://github.com/nextauthjs/next-auth/blob/58e30a6af6e8b1f9084f5edc83aa83ddf44653ec/packages/next-auth/src/core/index.ts#L166-L172

What should the rule for adding the `callbackUrl` on the custom page?

**Comment by ConnectWithNoor** [comment]
Is their any work around for the problem? I'm too stuck in the same issue

**Comment by mtrogers** [comment]
I am also running in to this. Navigating to /protected redirects to the customer signin page `http://localhost:3000/login?callbackUrl=%2Fprotected`, great. Seems as expected, if not exactly pretty.
But signing in blows it up. With` redirect: false`, 
```
 signIn("credentials", {
                    redirect: false,
                    email: e.currentTarget.email.value,
                    password: e.currentTarget.password.value,
                } ).then(({ok, error, url}) => {
                    console.log('okay, error: ', ok, error, url);
                    }
                });
```
Yields: {true, null, http://localhost:3000/login?callbackUrl=%2Fprotected}

I feel like i'm just missing a snippet line in the documentation somewhere...

**Reply by mtrogers** [reply]
I solved this, but in what feels like an unsatisfying way. In the form component where I called `signIn()`, I had to make some changes. I use the `useRouter()` and `userSearchParam()` hooks to get the callback url. I am concerned that this will allow malicious callback urls, but I guess I'll have to check for this. I then use `router.push()` to navigate to the previously attempted page. And of course, `e.preventDefault()`.

```
import {useRouter, useSearchParams} from "next/navigation";
import { useSession } from "next-auth/react";

export default function Form() {
...
  const router = useRouter();
  const query = useSearchParams();
           onSubmit={(e) => {
                e.preventDefault();
                setLoading(true);
                signIn("credentials", {
                    redirect: false,
                    email: e.currentTarget.email.value,
                    password: e.currentTarget.password.value,
              
                } ).then(({ok, error, url}) => {
                    setLoading(false);
                    if (error) {
                        toast.error(error as string);
                    }
                    else if (ok) {
                        const callbackUrl = query.get('callbackUrl')
                        router.push(callbackUrl || '/');
                    }
                });

```

**Reply by mtrogers** [reply]
Yup, if I navigate to the page http://localhost:3000/login?callbackUrl=https%3A%2F%2Ffacebook.com%0D%0A, it does take me to facebook after I login. Sigh. I may just have to accept navigation to a static page like `/dashboard`

**Reply by mtrogers** [reply]
Okay, I have a more workable solution. I can't get the benefit of the flicker-less page load of using router, but at least I can secure the callbackUrl.

```
...
signIn("credentials", {
                    // redirect: false, // default is false
                    email: e.currentTarget.email.value,
                    password: e.currentTarget.password.value,
                    callbackUrl: query.get('callbackUrl'));
```

**Reply by mtrogers** [reply]
Digging through the next-auth package, I found this in core/lib/callback-url.js:

```
async function createCallbackUrl({
  options,
  paramValue,
  cookieValue
}) {
  const {
    url,
    callbacks
  } = options;
  let callbackUrl = url.origin;

  if (paramValue) {
    callbackUrl = await callbacks.redirect({
      url: paramValue,
      baseUrl: url.origin
    });
  } else if (cookieValue) {
    callbackUrl = await callbacks.redirect({
      url: cookieValue,
      baseUrl: url.origin
    });
  }

  return {
    callbackUrl,
    callbackUrlCookie: callbackUrl !== cookieValue ? callbackUrl : undefined
  };
}
```

So attaching a debugger here and logging in at http://localhost:3000/login?callbackUrl=%2Fprotected, yields:

> paramValue = undefined
> cookieValue = "http://localhost:3000/login?callbackUrl=%2Fprotected"
> callbackUrl = "http://localhost:3000/login?callbackUrl=%2Fprotected"
> url.origin = "http://localhost:3000"

and

```
 return {
    callbackUrl,
    callbackUrlCookie: callbackUrl !== cookieValue ? callbackUrl : undefined
  };
```

So looks like we're passing to the redirect function specified in the documentation:

```
pages/api/auth/[...nextauth].js
callbacks: {
  async redirect({ url, baseUrl }) {
    // Allows relative callback URLs
    if (url.startsWith("/")) return `${baseUrl}${url}`
    // Allows callback URLs on the same origin
    else if (new URL(url).origin === baseUrl) return url
    return baseUrl
  }
}
```

It'll just end up with 

> url = "http://localhost:3000/login?callbackUrl=%2Fprotected"
> baseUrl = "http://localhost:3000"

And eventually return "http://localhost:3000/login?callbackUrl=%2Fprotected"
So, that's not going to help if you don't specify the `callbackUrl`. I wonder if I could access the route query here some how, or something. Probably not really worth it, better to just specify the `callbackUrl` option at the `signIn()` level.

**Reply by christo9090** [reply]
Thanks a ton. I was trying to work this out forever but this works for me. Seems like a pretty frustrating bug considering almost everyone is going to make a custom login page.

**Reply by abhishekslab** [reply]
facing this myself, this a very weird bug, I thought I was doing something wrong or missing something obvious, given this is such a normal flow for anyone who is really building auth

**Comment by svixxn** [comment]
I had the same issue, maybe you didn't defined your protected routes in middleware.ts file. If you don't do that, that will make all your routes  protected, so you won't be able to sign in and get in the infinite redirect loop. Check your middleware.ts:

```
export { default } from "next-auth/middleware";

export const config = { matcher: ["/dashboard"] };
```
in the matcher array specify your protected routes.

**Comment by josepmesteves** [comment]
I had the same issue of the callbackUrl growing every time I accessed the custom login page.

I solved it by building on top of what @mtrogers did on [his comment](https://github.com/nextauthjs/next-auth/discussions/3748#discussioncomment-5734560).

My solution keeps the original callback url that the user had when they landed on the custom login page. Here it is:

```
    async redirect({ url, baseUrl }: { url: string, baseUrl: string }) {
      const isRelativeUrl = url.startsWith("/");
      if (isRelativeUrl) {
        return `${baseUrl}${url}`;
      }

      const isSameOriginUrl = new URL(url).origin === baseUrl;
      const alreadyRedirected = url.includes('callbackUrl=')
      if (isSameOriginUrl && alreadyRedirected) {
        const originalCallbackUrl = decodeURIComponent(url.split('callbackUrl=')[1]);

        return originalCallbackUrl;
      }

      if (isSameOriginUrl) {
        return url;
      }

      return baseUrl;
    }
```

You should use this method inside the `callbacks` field that belongs to the auth options you're passing to NextAuth, on file `src/app/api/auth/[...nextauth]/route.ts` (or whatever's the equivalent you have).
